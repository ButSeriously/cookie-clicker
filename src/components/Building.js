import React from 'react';
import PropTypes from 'prop-types';
import smallCookie from '../img/perfectCookie-small.png';

class Building extends React.Component {
  state = {
    amount: 0,
    buttonClass: 'store_panel__building',
    cost: this.props.data.initialCost,
    cookiesPerSecond: this.props.data.productionPerSecond,
    isShown: false,
  }

  componentWillReceiveProps = () => {
    const hiddenButtonClass = 'store_panel__building store_panel__building--hidden';
    if (
      this.props.cookiesAmount < this.state.cost &&
      hiddenButtonClass !== this.state.buttonClass
    ) {
      // console.log('no');
      this.setState(() => ({
        buttonClass: hiddenButtonClass,
        isShown: false,
      }));
    } else if (
      this.props.cookiesAmount >= this.state.cost &&
      hiddenButtonClass === this.state.buttonClass
    ) {
      // console.log('yes');
      this.setState(() => ({
        buttonClass: 'store_panel__building',
        isShown: true,
      }));
    }
  }

  onMouseDown = () => {
    if (!this.state.isShown) return;
    this.setState(() => ({
      buttonClass: 'store_panel__building--clicked',
    }));
  }

  onMouseUp = () => {
    if (!this.state.isShown) return;
    this.setState(() => ({
      buttonClass: 'store_panel__building',
    }));
  }

  buyBuilding = () => {
    if (this.props.cookiesAmount < this.state.cost) return;
    this.setState(prevState => ({
      amount: prevState.amount + 1,
    }), () => {
      this.props.buildingBought(this.state.cost, this.state.cookiesPerSecond);
      this.increaseCost();
    });
  }

  increaseCost = () => {
    this.setState(prevState =>
      ({ cost: Math.floor(prevState.cost * 1.15) }));
  }

  render() {
    return (
      <button
        className={this.state.buttonClass}
        onClick={this.buyBuilding}
        onMouseDown={this.onMouseDown}
        onMouseUp={this.onMouseUp}
      >
        <div>
          <p className="store_panel__building_name">{this.props.data.name}</p>
          <p className="store_panel__building_cost">
            <img src={`./public/${smallCookie}`} alt="Cookie.png" />
            {` ${this.state.cost}`}
          </p>
        </div>
        <p className="store_panel__building_amount">{this.state.amount}</p>
      </button>
    );
  }
}

Building.propTypes = {
  data: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])).isRequired,
  cookiesAmount: PropTypes.number.isRequired,
  buildingBought: PropTypes.func.isRequired,
};

export default Building;
