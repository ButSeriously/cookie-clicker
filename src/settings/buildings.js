const buildings = [
  {
    name: 'Cursor',
    description: 'Autoclicks once every 10 seconds.',
    initialCost: 15,
    productionPerSecond: 0.1,
  }, {
    name: 'Grandma',
    description: 'A nice grandma to bake more cookies.',
    initialCost: 100,
    productionPerSecond: 1,
  }, {
    name: 'Farm',
    description: 'Grows cookie plants from cookie seeds.',
    initialCost: 1100,
    productionPerSecond: 8,
  }, {
    name: 'Mine',
    description: 'Mines out cookie dough and chocolate chips.',
    initialCost: 12000,
    productionPerSecond: 47,
  }, {
    name: 'Factory',
    description: 'Produces large quantities of cookies.',
    initialCost: 130000,
    productionPerSecond: 260,
  },
];

export default buildings;
