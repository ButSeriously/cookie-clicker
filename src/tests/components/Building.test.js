import React from 'react';
import { shallow } from 'enzyme';
import Building from '../../components/Building';
import buildings from '../../settings/buildings';

test('should render Building', () => {
  const wrapper = shallow(<Building data={buildings[0]} />);
  expect(wrapper).toMatchSnapshot();
});
