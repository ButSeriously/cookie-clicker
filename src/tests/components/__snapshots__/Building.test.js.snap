// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Building 1`] = `
<button
  className="store_panel__building"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
>
  <div>
    <p
      className="store_panel__building_name"
    >
      Cursor
    </p>
    <p
      className="store_panel__building_cost"
    >
      <img
        alt="Cookie.png"
        src="./public/test-file-stub"
      />
       15
    </p>
  </div>
  <p
    className="store_panel__building_amount"
  >
    0
  </p>
</button>
`;
